File: leetcode/algorithms/string/reverse-only-letters.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Reverse Only Letters"
*/

class Solution {
    public String reverseOnlyLetters(String S) {
        ArrayList<Integer> letterIndices = new ArrayList<Integer>();
        StringBuilder reverseBuilder = new StringBuilder(S.length());
        for(int i=0;i<S.length();++i){
            if(Character.isLetter(S.charAt(i))){
                letterIndices.add(i);
            }
            reverseBuilder.append(S.charAt(i));
        }
        
        for(int i=0;i<Math.floor(letterIndices.size()/2);++i){
            int thisIndex = letterIndices.get(i).intValue();
            int swapIndex = letterIndices.get((letterIndices.size()-i)-1).intValue();
            char temp = reverseBuilder.charAt(thisIndex);
            reverseBuilder.setCharAt(thisIndex,reverseBuilder.charAt(swapIndex));
            reverseBuilder.setCharAt(swapIndex,temp);
        }
        return reverseBuilder.toString();
        
    }
}

File: leetcode/algorithms/string/reverse-only-letters.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Reverse Only Letters"
*/

/**
 * @param {string} S
 * @return {string}
 */
var reverseOnlyLetters = function(S) {
    let low=0,high=S.length-1;
    
    const charArr = S.split('');
    
    while(high>low){
        if(!((S.charCodeAt(low)>64 && S.charCodeAt(low)<91)
           ||
           (S.charCodeAt(low)>96 && S.charCodeAt(low)<123))){
            charArr[low]=S[low];
             low++;
        } else if(!((S.charCodeAt(high)>64 && S.charCodeAt(high)<91)
           ||
           (S.charCodeAt(high)>96 && S.charCodeAt(high)<123))){
            charArr[high]=S[high];
             --high;   
        } else {
            charArr[low] = S[high];
            charArr[high] = S[low];
            ++low;
            --high;
        }
        
        
    }
    
    return charArr.join('');
    
};